File: CodeQuest 4.0 Jr-SY Round-1 Solutions/Mystical Realm of Eldoria.py
# taking input of the number of testcases
t = int(input())
for _ in range(t):
    n, h = map(int, input().split())

    # taking input for the attack strikes
    a = list(map(int, input().split()))

    # setting limits for binary search between maximum limit and minimum limit for the value of poison damage (according to constraints)
    l, r = 1, 10**18
    while l <= r:
        m = (l + r) // 2
        s = m
        for i in range(n - 1):
            s += min(m, a[i + 1] - a[i])
        if s < h:
            l = m + 1
        else:
            r = m - 1
    print(r + 1)

File: CodeQuest 4.0 Jr-SY Round-1 Solutions/Mystical Realm of Eldoria.cpp
#include <bits/stdc++.h>
using namespace std;
using li = long long; // long long integers are used for avoiding overflow errors
int main() {
  int t; // taking input of the number of testcases
  cin >> t;
  while (t--) {
    int n;
    li h;
    cin >> n >> h;

    // taking input for the attack strikes
    vector<li> a(n);
    for (li &x : a) cin >> x;

    // setting limits for binary search between maximum limit and minimum limit for the value of poison damage (according to constraints)
    li l = 1, r = 1e18;
    while (l <= r) {
      li m = (l + r) / 2;
      li sum = m;
      for (int i = 0; i < n - 1; ++i) 
        sum += min(m, a[i + 1] - a[i]);
      if (sum < h) l = m + 1;
      else r = m - 1;
    }
    cout << r + 1 << '\n';
  }
}